Wikipedia/Code/WMFArticleViewController.m

```diff
@@ -1586,6 +1586,11 @@ - (void)webViewController:(WebViewController *)controller didScrollToSection:(MW
     self.visibleSectionAnchor = section.anchor;
 }
 
+- (void)webViewController:(WebViewController *)controller didScrollToFragment:(NSString *)fragment {
+    self.visibleSectionAnchor = fragment;
+    [self updateTOCHighlightIfNecessaryWithScrollView:controller.webView.scrollView force:true];
```

Comment:
Any reason for not selecting the correct section based on the fragment directly? ```
     if (!fragment) {
        return;
    }
    MWKSection *section = [self.article.sections sectionWithFragment:fragment];
    [self selectAndScrollToTableOfContentsItemForSection:section animated:NO]; 
``` with this change and the delay removed, it seems to fix the issue on click through and on the color article.

Reply:
I pushed this change because I saw this PR is tagged with changes welcome but feel free to revert or modify if there was something I missed.

Reply:
Not all fragments are sections.

Reply:
At least, it seemed safer not to assume so...

Reply:
If that's the case, it might make sense to fall back on the old method if the section returned is `nil`